packages/sit-onyx/src/composables/useCustomValidity.ts

```diff
@@ -61,34 +83,88 @@ export const useCustomValidity = (options: UseCustomValidityOptions) => {
        */
       watchEffect(() => el.setCustomValidity(options.props.customError ?? ""));
 
-      /**
-       * Update validityState ref when the input changes.
-       */
       watch(
-        [() => options.props.customError, () => options.props.modelValue, isTouched],
+        [() => options.props.customError, () => options.props.modelValue],
         () => {
           const newValidityState = transformValidityStateToObject(el.validity);
 
-          // do not emit validityChange event if input is valid and has never been invalid
-          if (!isTouched.value || (!validityState.value && newValidityState.valid)) return;
+          // do not emit update if input is valid and has never been invalid
+          if (!validityState.value && newValidityState.valid) return;
 
           // ignore if actual validity state value is unchanged
           if (validityState.value && areObjectsFlatEqual(newValidityState, validityState.value)) {
             return;
           }
 
           validityState.value = newValidityState;
+        },
+        { immediate: true },
+      );
+
+      /**
+       * Update validityState ref when the input changes.
+       */
+      watch(
+        [() => options.props.customError, validityState, isDirty],
+        () => {
+          // do not emit validityChange event if the value was never changed
+          if (!isDirty.value || !validityState.value) return;
+
           options.emit("validityChange", validityState.value);
         },
         { immediate: true },
       );
     },
   } satisfies Directive<InputValidationElement, undefined>;
 
+  const errorMessages = computed<{ longMessage?: string; shortMessage: string }>(() => {
+    if (!validityState.value || validityState.value.valid) return { shortMessage: "" };
+
+    const errorType = getFirstInvalidType(validityState.value);
+    // a custom error message always is considered first
+    if (options.props.customError || errorType === "customError") {
+      const message = options.props.customError ?? "";
+      return { shortMessage: message, longMessage: message };
+    }
+    if (!errorType) return { shortMessage: "" };
+
+    // if the error is "typeMismatch", we will use an error message depending on the type property
+    if (errorType === "typeMismatch") {
+      const type = TRANSLATED_INPUT_TYPES.includes(options.props.type as TranslatedInputType)
+        ? (options.props.type as TranslatedInputType)
+        : "generic";
+      return {
+        longMessage: t.value(`validations.typeMismatch.${type}`, {
+          value: options.props.modelValue?.toString(),
+        }),
+        shortMessage: t.value(`validations.preview.typeMismatch.${type}`, {
+          value: options.props.modelValue?.toString(),
+        }),
+      };
+    }
+
+    const validationData = {
+      value: options.props.modelValue?.toString(),
+      n: options.props.modelValue?.toString().length ?? 0,
```

Comment:
`.toString()` works for input and textarea but not for selects because its an array there...maybe we should change the type of modelValue to `string` for now?

Reply:
then we need to convert to string when we call the composable, right?

Reply:
Not for now since the type is only string for input and textarea, when supporting the select in the future, we can maybe pass in the displayed string value instead

Reply:
the composable is also used in checkbox, switch, ... where the value can be boolean.
select is the next step and part of this ticket #570

Reply:
Ah okay, I guess its fine for now then